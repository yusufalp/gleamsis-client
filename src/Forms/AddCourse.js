import React from 'react';
import { Link } from 'react-router-dom';
import DataContext from '../DataContext';

class AddCourse extends React.Component {
  static contextType = DataContext
  state = {
    name: '',
    category: '',
    teacherId: '',
    error: ''
  }
  handleCourseName(value) {
    this.setState({
      name: value.trim()
    })
  }
  handleCourseCategory(value) {
    this.setState({
      category: value
    })
  }
  handleTeacherId(value) {
    this.setState({
      teacherId: value
    })
  }
  handleCourseSubmit(e) {
    e.preventDefault()

    if(!this.state.name || !this.state.category || !this.state.teacherId){
      this.setState({
        error : "Name, category and teacher are required"
        })
        return
    }

    let newCourse = {
      id: this.context.courses.length + 1,
      name: this.state.name,
      category: this.state.category,
      teacher_id: Number(this.state.teacherId)
    }

    // add newCourse to the context
    this.context.addCourse(newCourse)

    this.props.history.push('/success')
  }
  render() {
    return (
      <div>
        <h2>Add a Course</h2>
        <p>
          Use the form below to add a course. You must add a course if there is a teacher to teach it.
          If the teacher does not exist, <Link to={'/AddTeacher'}>add the teacher</Link> first.
      </p>
        <form onSubmit={e => this.handleCourseSubmit(e)}>
          <label htmlFor='name'>Course Name*</label>
          <input
            type='text'
            name='name'
            id='name'
            onChange={e => this.handleCourseName(e.target.value)}
            required
          />
          <label htmlFor='category'>Category*</label>
          <input
            type='text'
            name='category'
            id='category'
            onChange={e => this.handleCourseCategory(e.target.value)}
            required
          />
          <label htmlFor='teacher'>Teacher*</label>
          <select
            name='teacher'
            id='teacher'
            required
            onChange={e => this.handleTeacherId(e.target.value)}>
            <option value="">Select a Teacher</option>
            {this.context.teachers.map((teacher, i) =>
              <option key={i} value={teacher.id}>{teacher.firstName} {teacher.lastName}</option>
            )}
          </select>
          <button type="submit">Add Course</button>
          <p>{this.state.error}</p>
          <p className="required">* = required fields</p>
        </form>
      </div>
    )
  }
}

export default AddCourse;