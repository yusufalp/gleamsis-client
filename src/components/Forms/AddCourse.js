import React from 'react';
import { Link } from 'react-router-dom';
import DataContext from '../DataContext';

class AddCourse extends React.Component {
  static contextType = DataContext;
  state = {
    name: '',
    category: '',
    teacherId: '',
    error: ''
  }
  handleCourseName(value) {
    this.setState({
      name: value.trim()
    })
  }
  handleCourseCategory(value) {
    this.setState({
      category: value.trim()
    })
  }
  handleTeacherId(value) {
    this.setState({
      teacherId: value
    })
  }
  handleCourseSubmit(e) {
    e.preventDefault();

    // Checks if the required information is entered to create a new course
    if (!this.state.name || !this.state.category || !this.state.teacherId) {
      this.setState({
        error: 'Name, category and teacher are required'
      })
      return;
    }

    let newCourse = {
      name: this.state.name,
      category: this.state.category,
      teacher_id: Number(this.state.teacherId)
    }

    this.context.addCourse(newCourse);
  }
  render() {
    return (
      <div>
        <header>
          <h2>Add a Course</h2>
        </header>
        <p className='add-info'>
          Use the form below to add a course. You must add a course if there is a teacher to teach it.
          If the teacher does not exist, <Link to={'/add-teacher'} className='form-link'>add the teacher</Link> first.
      </p>
        <form onSubmit={e => this.handleCourseSubmit(e)}>
          <label htmlFor='name'></label>
          <input
            type='text'
            name='name'
            id='name'
            onChange={e => this.handleCourseName(e.target.value)}
            placeholder='Course Name'
            required
          />
          <label htmlFor='category'></label>
          <input
            type='text'
            name='category'
            id='category'
            onChange={e => this.handleCourseCategory(e.target.value)}
            placeholder='Category'
            required
          />
          <label htmlFor='teacher'></label>
          <select
            name='teacher'
            id='teacher'
            required
            onChange={e => this.handleTeacherId(e.target.value)}>
            <option value=''>Select a Teacher</option>
            {this.context.teachers.map((teacher, i) =>
              <option key={i} value={teacher.id}>{teacher.first_name} {teacher.last_name}</option>
            )}
          </select>
          <button type='submit'>Add Course</button>
          <p>{this.state.error}</p>
          <p className='required'>* = required fields</p>
        </form>
      </div>
    )
  }
}

export default AddCourse;