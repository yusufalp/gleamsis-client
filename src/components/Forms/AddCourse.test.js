import React from 'react';
import ReactDOM from 'react-dom';
import AddCourse from './AddCourse';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

describe('AddCourse component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <AddCourse />
      </MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <AddCourse />
        </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});