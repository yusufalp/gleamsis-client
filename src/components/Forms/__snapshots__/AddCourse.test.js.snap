// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCourse component renders the UI as expected 1`] = `
<div>
  <header>
    <h2>
      Add a Course
    </h2>
  </header>
  <p
    className="add-info"
  >
    Use the form below to add a course. You must add a course if there is a teacher to teach it. If the teacher does not exist, 
    <a
      className="form-link"
      href="/add-teacher"
      onClick={[Function]}
    >
      add the teacher
    </a>
     first.
  </p>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="name"
    />
    <input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Course Name"
      required={true}
      type="text"
    />
    <label
      htmlFor="category"
    />
    <input
      id="category"
      name="category"
      onChange={[Function]}
      placeholder="Category"
      required={true}
      type="text"
    />
    <label
      htmlFor="teacher"
    />
    <select
      id="teacher"
      name="teacher"
      onChange={[Function]}
      required={true}
    >
      <option
        value=""
      >
        Select a Teacher
      </option>
    </select>
    <button
      type="submit"
    >
      Add Course
    </button>
    <p>
      
    </p>
    <p
      className="required"
    >
      * = required fields
    </p>
  </form>
</div>
`;
